### ISSUE TITLE
Fetch body streams are not full duplex

### ISSUE THREAD

**Issue by lucacasonato** [whatwg/fetch#1254]
With request body streams and response body streams one would think it is now possible to use `fetch` for full duplex communication over HTTP, like in the example below:

```js
const intervalStream = new ReadableStream({
  start(c) {
    let count = 0;
    const timer = setInterval(() => {
      console.log("sent!");
      c.enqueue("Hello\n");
      if (count === 5) {
        clearInterval(timer);
        c.close();
      }
      count++;
    }, 1000);
  },
}).pipeThrough(new TextEncoderStream());

const resp = await fetch("https://full-duplex-server.deno.dev/", {
  method: "POST",
  body: intervalStream,
});

console.log("foo");

const reader = resp.body.pipeThrough(new TextDecoderStream()).getReader();

while (true) {
  const { value, done } = await reader.read();
  if (done) break;
  console.log(value);
}

console.log("done!");
```

In this example the server just pipes the incoming request body stream into the response body steam with no modification.

If full duplex communication was working,  you would see "sent!" and "Hello\n" being logged in close proximity to each other, live. This is however not the case, because the promise returned from `fetch` does not resolve until the entire request body stream has been sent and closed.

The solution here would be to resolve the `Promise<Response>` returned from `fetch` before the request body stream has completed. To do this, body upload streaming has to happen in parallel. The main issue with the spec in this area is that the promise returned from `fetch` is expected to reject if a chunk of the request body stream is not a Uint8Array (behaviour is tested in https://staging.wpt.fyi/results/fetch/api/basic/request-upload.any.html).

Are there any plans to resolve this shortcoming before request body streaming is sthbilized in browsers? HTTP/2 and QUIC are very capable duplex stream protocols (HTTP/1 is theoretically also capable) and it would be a shame if that power was not exposed to the web platform.

**Comment by lucacasonato** [comment]
To be clear, I am not explicitly vouching for duplex streams to make it into the initial request body uploading implementation in browsers, I just want to make sure that it would be possible to add down the line. Deno implements `fetch` w/ request body streaming, and our implementation is capable of full duplex too, but it has to deviate from the spec to allow that. The backing HTTP implementations for both Node.js and Servo are also capable of full duplex body streams.

**Comment by annevk** [comment]
See #229 for some history. Since then, I've learned that this would require a lot of work in browsers and it's not planned to be supported initially.

However, I do think we should allow for it in theory, so if you could elaborate on what prohibits it that would help.

**Comment by lucacasonato** [comment]
> so if you could elaborate on what prohibits it that would help

`fetch` does not complete until the entire ReadableStream has been read and uploaded. This is fundamentally incompatible with duplex streaming, as you can not get a handle to the `response.body` until you close `request.body`.

These tests demonstrate that issue: https://github.com/web-platform-tests/wpt/blob/master/fetch/api/basic/request-upload.any.js. It is expected that the promise returned from `fetch` rejects if a chunk that is pulled off the `request.body` is not a `Uint8Array`.

**Comment by annevk** [comment]
Does the specification also have such issues?

Now that you mention that example though I realize it is rather tricky if not impossible to do well here and be forward compatible. I suspect that applications will have the same kind of dependencies and browsers changing this will break them.

**Comment by sleevi** [comment]
Yeah, there seems some overlap with #95 / #538 in terms of upload/request processing.

**Comment by annevk** [comment]
cc @yoichio @yutakahirano

**Comment by lucacasonato** [comment]
> Does the specification also have such issues?

As far as I can tell, yes. The request body is sent synchronously before receiving response headers (not in parallel). See the "To transmit request’s body body, run these steps:" section in [HTTP-network fetch](https://fetch.spec.whatwg.org/#http-network-fetch) step 8.5. Essentially what needs to happen is, that this needs to be changed to happen in parallel with receiving the response headers.

> Now that you mention that example though I realize it is rather tricky if not impossible to do well here and be forward compatible.

To elaborate on this: many (likely most) web servers are programmed to only send back response headers after the request body has been received and processed (from now on referred to as category one servers). Because the promise returned from `fetch` would only resolve once the response headers are received. For the user the behaviour is that "fetch completes after the request body is fully sent". This behaviour would not change once full duplex streaming is implemented in browsers because the response headers are not received by the browser until the request body has been sent (because only then are they sent by the server).

For servers that send back response headers before the request body has been received (from now on referred to as category two servers), the promise returned from `fetch` would still only resolve once the request body is fully sent (because as you say, browsers do not support full duplex HTTP right now). Once browsers implement full duplex HTTP the promise returned from `fetch` would resolve after the response headers are received, regardless of the state of the request body. This is technically a minor change in behaviour, but one that could be marked as a spec compliance bug that Chrome, FF, and Safari have. The Deno or Servo implementations could implement full duplex right away.

Because the Python WPT test server probably falls into category one, the WPT I mentioned above is technically correct, but only under the assumption that the server sends headers after the response body is received. For cases where you need to handle both category one and category two servers you should handle the "cancellation" of the `ReadableStream`. This would be the only way to get notified of request body errors for category two servers.

--- 

TLDR, to solve this following would need to be resolved in the spec:

- [ ] Request body sending in _HTTP-network fetch_ should happen in parallel to waiting for response body headers.

**Comment by lucacasonato** [comment]
One question this brings up: if request body sending on _HTTP-network fetch_ happens in parallel, how and where are errors propagated to the user?

**Comment by annevk** [comment]
In the case where the caller is not using upload streams? I suppose they would need `FetchObserver` (#607).

**Comment by lucacasonato** [comment]
Yeah. For users that are using upload streams it can be surfaced as a cancellation of the stream.

Maybe the error could be surfaced as a rejection of the promise returned by `fetch` if the response headers have not yet been received, or alternatively as an error on the response body stream. Sort of like how aborting fetch works, but with a custom error message (the error that request body upload errored with), instead of an `AbortError`.

This would introduce the constraint that request body uploading must not outlive response body download. Good chance that is already a HTTP spec constraint anyway, but sure though.

**Comment by ricea** [comment]
My view:
* HTTP is a half-duplex request/response protocol and should remain so
* WebSocket or [WebTransport](https://www.w3.org/TR/webtransport/) should be used where full-duplex is required. In particular, they have mitigations against request smuggling.
* My team would be responsible for implementing this in Chromium, and even if we decided we wanted to do it, we don't have the resources to do the necessary rearchitecting of the network stack. Even just upload streaming has proven more challenging than expected.
* The Fetch Standard should continue to reflect what browsers actually implement.

**Comment by annevk** [comment]
@ricea the assertion that HTTP is half-duplex seems in contradiction with it allowing the client to consume the response before it has transmitted the request body. I tend to agree with regards to the cost/benefit tradeoff and what Fetch should specify (although I would like for things to be cooler).

**Comment by lucacasonato** [comment]
@ricea Some comments:

HTTP is not half duplex. It allows for simultaneous request body and response body streaming at a per request level. This is supported in both HTTP/1.1. and HTTP/2. Here are some discussions around the topic from 8 years ago: https://datatracker.ietf.org/doc/html/draft-zhu-http-fullduplex#section-2.1. Note that this is full duplex on a **per request level**, not full duplex on a connection level (the latter is only supported in HTTP/2, and browsers make ample use of it). Go, Node.js, and many other runtimes support full-duplex HTTP body streams.

Regarding request smuggling: full-duplex is not any more vulnerable to this than any other request body streaming (#966). Whatever the ultimate restrictions are that result from the #966 discussion, they would apply to full-duplex request body streaming too.

Regarding the effort required for implementation: it is very understandable that there are no resources to implement this now. My concern is that with the current spec wording there is also no way to implement it in the future, even if opinion on this matter changes. We are effectively locked into half-duplex for fetch forever. I think the specification should leave room for specifying this behaviour in the future - who knows what happens 5 years from now :-)

Also some non-browser engines implement the Fetch specification now, or are planning to in the future (Deno, Node.js, Cloudflare Workers). I think it is worth discussing if the specification should cater to these audiences too - I think APIs being the same across browsers and servers is a great thing that nearly no other language can provide. Also the Servo network stack is already capable of full-duplex HTTP.

**Comment by domenic** [comment]
> Regarding the effort required for implementation: it is very understandable that there are no resources to implement this now. My concern is that with the current spec wording there is also no way to implement it in the future, even if opinion on this matter changes. We are effectively locked into half-duplex for fetch forever. I think the specification should leave room for specifying this behaviour in the future - who knows what happens 5 years from now :-)

That's not really how specifications work (at least in the WHATWG). Implementations aren't prohibited from implementing things because the spec doesn't support them. Instead, implementations go ahead and implement ahead of or in parallel with spec work, and then bring their findings back to the spec for discussion. If there's multiple implementers interested in shipping something, then we change the spec, around the same time as it ships.

So I don't think you should be worried about the fact that the spec reflects implementations. That's working as intended. The spec can change, if resources to implement this free up.

**Comment by ioquatix** [comment]
Having implemented full-duplex HTTP clients and servers for Ruby, I fully support @lucacasonato's position and think we should guide the fetch spec towards being full duplex.

If browsers can't support it, that's understandable, but if someone wants to implement a browser (or client) that *does*, it should not be prevented from being compatible with the spec here IMHO.

**Comment by benbucksch** [comment]
> [Problem:] the promise returned from fetch does not resolve until the entire request body stream has been sent and closed.
> The solution here would be to resolve the Promise<Response> returned from fetch before the request body stream has completed

Agreed. That would also be the expected behavior for on-duplex upload-only. I want to know whether the connection is properly established, the promise to resolve, and then start streaming. I.e. I would expect the promise to resolve in the exact moment when the upload stream can start.

**Comment by benbucksch** [comment]
For download streams, the connection and streaming body are 2 separate steps, which can be awaited separately. The same would make sense for upload.

**Comment by annevk** [comment]
To be clear, I don't think we need to resolve the promise returned from `fetch()` any sooner. Resolving it once we have the response headers in is fine. That's the signal. If you care about the request you monitor its stream.

**Comment by benbucksch** [comment]
> [not] upload stream racing with the response headers

Yes, indeed, I would expect the response headers to have returned before the fetch() promise returns. Once the fetch() promise returns, I would immediately start sending stream data. That's a simple and logical API.

**Comment by annevk** [comment]
Now that #1457 has merged this becomes a feature request.

**Comment by lucacasonato** [comment]
FYI, Chrome only shipping half duplex has caused definitely user confusion. Users are expecting full duplex (without knowing that is what it is called), and as such even when specifying `duplex: "half"` (without knowing why, Chrome just says you have to add it) they do not understand that they are only getting half duplex.

For example, this Deno user expected full duplex streaming and was disappointed / confused as to why only half duplex is working: https://www.reddit.com/r/Deno/comments/150wet7/why_doesnt_servetls_stream_data_from_a/

@yutakahirano Is Chrome open to reconsider its stance on implementing full duplex if we do the spec work & WPT tests?

**Comment by lucacasonato** [comment]
I'd also be interested to hear WebKit (@annevk) and Gecko (@valenting)

**Comment by ioquatix** [comment]
> Chrome only shipping half duplex has caused definitely user confusion

I'm honestly not surprised by this, and I would have expected full duplex by default, and I support making this full duplex by default if possible.

**Comment by ricea** [comment]
> @yutakahirano Is Chrome open to reconsider its stance on implementing full duplex if we do the spec work & WPT tests?

Yutaka no longer works on Chrome. My response is unchanged from https://github.com/whatwg/fetch/issues/1254#issuecomment-862810798.

**Comment by ricea** [comment]
The response is not sent until the request (including body) has been completely received. That's what I mean by "half-duplex".

Smart people have been known to disagree about what the HTTP RFCs say on this point, so I am going to stay out of that rabbit hole.

**Comment by ioquatix** [comment]
Do you mind pointing us at the statements regarding "half-duplex" in the RFCs that "Smart people have been known to disagree about"?

**Comment by annevk** [comment]
@ioquatix I don't think that inquiry is a productive use of everyone's time. Fact is that in all browser engines full duplex would be a substantial undertaking and alternatives such as WebSockets and soon WebTransport are available. The benefits and demand for full duplex HTTP would have to be quantified quite a bit to convince anyone I suspect.

**Comment by ioquatix** [comment]
What level of quantification is required? Because there are quite a few people on this thread who are already convinced it's a requirement.

If you ask for quantification, but that bar impossibly high or ambiguous to the point where it cannot be satisfied, that's not a productive use of anyone's time either.

If the point is "we standardise what the browsers do" and "my view (chrome) is that HTTP is half-duplex", then we are basically at an impasse as far as I can tell.

I have no problem championing a feature that I think is incredibly useful and important, but if the answer is simply "we don't want to do it for reasons", then let's not go down the path of trying to "quantify quite a bit to convince anyone".

On the other hand, if this is a genuine call to quantify whether such a feature is required, let's set the bar appropriately and then make a binding decision when appropriate evidence is collected.

**Comment by piranna** [comment]
When HTTP was being used just for request-response, It made sense of being half-duplex: made a (small) request, finish to fully receive It, process, and send response back. But now we can send streams, not just "fixed size" requests, so not only for performance reasons make sense to start processing the request without fully receiving It, but also send back the response, but also from an users and developer PoV makes sense too. Servers are already working like that for performance reasons (for example Node.js), missing part to standarize It are browsers. I know there can be problems if an issue appears in the middle of receiving the requests and the response started to be send, that's the kind of details the spec needs to solve, but for the "happy path", technically IS totally feasable to implement.

**Comment by annevk** [comment]
@ioquatix I don't speak for Chrome. I sometimes speak for WebKit (and in particular Apple's port), and used to sometimes speak for Gecko, but I'm also the editor of the Fetch Standard and responsible for this repository.

https://whatwg.org/working-mode#changes has the requirements for changing a WHATWG Standard. You will need implementer support. In the past, web developer demand for certain features as well as end user benefit has been able to sway implementers. No guarantees for the future, but it seems likely that still holds.

So how would you go about quantifying something? Showing lots of questions on Stack Overflow, existence of JS libraries that paper around the problem using WebSockets or WebTransport, demonstrating deficiencies in workarounds, etc. This can take a long time, doesn't always work out, and even when there is reasonable demand it might still not be enough to persuade all relevant parties. But it has generally worked which is why we keep issues requesting new features open.

**Comment by ioquatix** [comment]
Okay, that's a fair enough comment. Is it okay if I start collecting references to those things on this thread? e.g.

- https://stackoverflow.com/questions/76564358/full-duplex-communication-over-the-same-http-2-request-response-stream

Or should I document it elsewhere?

**Comment by annevk** [comment]
Here is fine, but maybe continue updating that comment so everyone watching the repository doesn't get a stream of small updates.

**Comment by helmturner** [comment]
@ioquatix I'd like to add to that. I recently started contributing to tRPC and, in the implementation of streaming functionality (in particularly high demand thanks to ChatGPT) is being partially blocked by this.

The use case for us was the ability to batch requests that include streamed responses. In order to cancel a single stream without cancelling the entire batch we need to get a message from server to client mid-response, potentially when that server is a stateless 'serverless' function. We already eagerly write blank Headers, and the intent would have been to use Trailers where necessary.

Without this, users are forced to pick between either batching OR streaming, or they have to implement some pretty painstaking request orchestration.

Re: 'Implementer support'... huh? As was already mentioned many times, server runtimes are implementing this already. In fact, it wasn't until our group read the spec, built a POC in node, AND did extra in-browser testing that we found out this was a no-go.

Finally, I want to add to the long list of obvious use-cases: simple push-messaging in React Native apps, without being forced to use Messaging APN providers (*cough* Firebase).

**Comment by benlesh** [comment]
Commenting to illustrate need:

I have a direct need for this as well. Working on a private code base, but the use case is we have a lot of streaming that we're currently multiplexing over a WebSocket. Most of these streams are simply "1 request, N responses", however 2 or 3 of them are bidirectional. Moving to http/2 for "1 request, N responses" case would be a boon for us, because we'd be able to let http/2 do the multiplexing for us, less JavaScript, less code running on the server and the client (because there's no need to wrap something in an "envelope" and match it up on both sides). The lack of bidirectional communication (full duplex) in fetch is really the only thing holding me up from moving everything over to http/2.

There's always `WebTransport` of course, but I feel like http/3 is an even taller order, I don't think it's supported in Node yet, and it's unlikely that the support is great in other languages our teams deal with.

**Comment by dojyorin** [comment]
There are many reasons why this is necessary, but I think it's especially important on corporate intranets.

Most of the time, such networks have (unnecessary) limitations on protocols like UDP/WebTransport and WebSocket that provide full-duplex communication.

The protocol itself for full-duplex fetch is just plain old HTTP, so it can provide full-duplex communication while avoiding these limitations.

**Comment by AjaniBilby** [comment]
Sorry to be that guy, but I just think it would be nice if I could start responding with data from the server while data from the client is still in-flight for easy UI updates is client libraries like htmx.

The lack of proper duplex support is literally the only thing preventing me from doing nice UI updates while a file is uploaded and ingested live on a single connection - and if that connection died before the upload is completed it can easily be detected and tasks can be cancelled and/or rolledback.

I understand this would be one hell of an undertaking for most browsers, but it would be nice if there at least an agreed goal towards this aiming vaugely at some point in the distant future.

**Comment by otyebis** [comment]
Its really frustrating that i have to use some ugly workarounds like always send two requests in a row with some identifiers to synchronize their use as input/output pair in order to keep the data flow. It feels like if i would be forced IRL to bring two bulky walkie-talkies everywhere i go instead of a compact rational portable telephone which was invented long ago and widely used. Its infuriating that its year 2025 and such a rational and semantically fitting core feature which would improve quality of life and code for any developer is still only on the stage of discussing. I demand urgent implementation.

**Comment by piranna** [comment]
@otyebis i'm also infuriated why such a basic feature is not a available yet, when it's pretty simple to achieve with any TCP socket, but if you would have read the other comments, you'll see it's NOT so simple to implement, since web browsers network libraries have been optimized since 30 years ago for the half-duplex request-response scheme, and modify them to support full-duplex bi-directional communication would means a massive architecture change for them, so don't hope to have It solved soon.

Also, spec writers are sometimes doing that in their spare time, and they doesn't own you anything, so relax your tone, because that's not constructive and only make people to don't employ time to work for free for people writting messages like yours that doesn't deserve It. And by the way, creating an account just only to write that message, it's of very bad taste and total lack of etiquete.

**Comment by Hashbrown777** [comment]
@AjaniBilby 
> Sorry to be that guy, but I just think it would be nice if I could start responding with data from the server while data from the client is still in-flight for easy UI updates is client libraries like htmx.
> 
> The lack of proper duplex support is literally the only thing preventing me from doing nice UI updates while a file is uploaded and ingested live on a single connection - and if that connection died before the upload is completed it can easily be detected and tasks can be cancelled and/or rolledback.
> 
> I understand this would be one hell of an undertaking for most browsers, but it would be nice if there at least an agreed goal towards this aiming vaugely at some point in the distant future.

You dont need duplex to do this, you dont need the server to tell the client how much of a file is uploaded, the client is doing the uploading.

I've achieved this before by using [TransformStream](https://developer.mozilla.org/en-US/docs/Web/API/TransformStream). Basically you just get your `File` from the form or drag or filepicker, then call `.stream().pipeThrough()` your custom `.transform()` which simply passes on all chunks untouched, but reports to the UI how many bytes it has seen. This stream is what you pass to the `fetch()` and voila, you have an accurate representation of bytes being ingested by the server (because the stream would block on your end if the server blocks on it end and stops or slows in receiving data). If there are followup processing steps you want monitored then followup with another http request to receive that information.

I can create a demo if you'd like. To others; fetch still needs to be duplex, there are other examples here that require it (I even need it, that's why I've been subscribed to this issue this whole time), this is just to help him out if he really needs that functionality, it's already been possible.

**Comment by AjaniBilby** [comment]
@Hashbrown777, I need information from the server - not just bytes transmitted.

For very large files (i.e. multi-gigabyte file ingesting) I want the server to be able to start responding immeditately since it ingests live because some files are literally too big to fit into cheap server rma.
Since the server is ingesting live it can respond with the pass rate of lines in a spreadsheet. That way if there is a really high error rate you can cancel the upload rather than having to let it finish see any results.

Right now the only way to achieve this would be to write my own desktop application just to make a single http request.
And then I also need to implement some client side rendering instead of just leaning on the browser.

**Comment by Hashbrown777** [comment]
> Right now the only way to achieve this would be to write my own desktop application

It definitely wouldnt be the only way if you're developing the backend too. In which case you could split the request into two tandem async fetch's for the upload and response data.
But yes in that scenario it would require actual engineering whichever way you choose (whether that's bespoke client app for duplex http or bespoke server code for syncing/relaying between the two requests) instead of my thinking you might've just wanted a progress bar for the upload (where the above would just drop in).

**Comment by AjaniBilby** [comment]
@Hashbrown777 two tandum requests sounds easy until you start counting the edge cases:

1. **Request failure**: If one or other request closes for some reason, do cancel the other? Attempt a restart?
2. **Fine load balancer control**: You need to make sure both requests end up on the same server, or have to use some state management like a DB to communicate across machines with live updates.
3. **Context Alignment**: You need to make sure not only both requests are executing on the same device/VM, but also they are in the same context. In some langauges and frameworks if the two requests end up on a different thread of execution this will add extreme amounts of overhead (i.e. ASP/PHP).

The only current work around would be to put the request through a binary websocket.

Also it's really not that bespoke. You only need about [~200 lines](https://github.com/AjaniBilby/recommend-moe/blob/main/app/util/form.ts#L8-L188) to write a streamed `multipart/form-data` iterator. Then another [150 lines](https://github.com/AjaniBilby/recommend-moe/blob/main/app/util/format/csv.ts#L43) to write a csv iterator that can ingest the form value chunk stream. Of course the examples I gave are JS, but even if you `10x` them into a real langauge, 2000 lines isn't much.

I'm honestly just surprised browsers aren't already capable of this, since they can render incomplete html as it's being streamed down the wire. I would have thought that would be a harder feature. But I guess that shows I don't know what network optimsations are going on under the hood.
